test/libyul/Parser.cpp

```diff
@@ -767,6 +769,102 @@ BOOST_AUTO_TEST_CASE(customSourceLocations_two_locations_with_snippets_untermina
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	std::shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result);
+	BOOST_REQUIRE(errorList.size() == 1);
+	BOOST_TEST(errorList[0]->type() == Error::Type::SyntaxError);
+	BOOST_TEST(errorList[0]->errorId() == 1544_error);
+	CHECK_LOCATION(result->debugData->originLocation, "", -1, -1);
+}
+
+BOOST_AUTO_TEST_CASE(customSourceLocations_two_snippets_with_hex_comment)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = R"(
+		/// @src 0:111:222 hex"abc"@src 1:333:444 "abc"
+		{}
+	)";
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	std::shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result && errorList.size() == 0);
+	// the second source location is not parsed as such, as the hex string isn't interpreted as snippet but
+	// as the beginning of the tail in AsmParser
+	CHECK_LOCATION(result->debugData->originLocation, "source0", 111, 222);
+}
+
+BOOST_AUTO_TEST_CASE(customSourceLocations_invalid_escapes)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = R"(
+		/// @src 0:111:222 "\n\\x\x\w\uö\xy\z\y\fq"
+		{}
+	)";
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	std::shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result && errorList.size() == 0);
+	CHECK_LOCATION(result->debugData->originLocation, "source0", 111, 222);
+}
+
+BOOST_AUTO_TEST_CASE(customSourceLocations_single_quote_snippet_with_whitespaces_and_escapes)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = R"(
+		/// @src 0:111:222 '\n\\x\x\w\uö\xy\z\y\fq'
+		/// @src 1 :		222 : 333 '\x33\u1234\t\n'
+		{}
+	)";
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	std::shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result && errorList.size() == 0);
+	CHECK_LOCATION(result->debugData->originLocation, "source1", 222, 333);
+}
+
+BOOST_DATA_TEST_CASE(customSourceLocations_scanner_errors_outside_string_lits_are_ignored, boost::unit_test::data::make({"0x ", "/** unterminated comment", "1_23_4"}), invalid)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = fmt::format(R"(
+		/// @src 0:111:222 {}
+		/// @src 1:222:333
+		{{}}
+	)", invalid);
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	std::shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result && errorList.empty());
+	CHECK_LOCATION(result->debugData->originLocation, "source1", 222, 333);
+}
+
+BOOST_AUTO_TEST_CASE(customSourceLocations_multi_line_source_loc)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = R"(
+		/// @src 1	: 111:
+		/// 222 "
+		/// abc\"def
+		///
+		/// " @src 0:333:444
```

Comment:
Where and how does this actually happen now - and how does it prevent
```
/// @src 1:1:1 "
let x := 42
/// "
```
from swallowing the non-comment?

Reply:
in the example it'll just skill all escaped stuff, so the snippet amounts to `abcdef`. the non-comment is not swallowed because it will never reach that part of the asmparser, there's an outer loop taking care of that, so only `1:1:1 "` would arrive and error out with unterminated quote

Reply:
so more precisely the input to `parseSrcComment` in the test you referenced up there is (with escapes)

`1\t: 111:\n222 "\nabc\"def\n\n" @src 0:333:444`

for which `(1,111,222)` is identified as components and `abcdef` as snippet (the rest is now skipped by scanner as discussed); everything else goes into the `tail` (being `@src 0:333:444`) which is returned to an outer loop.

The outer loop will match it as source location comment and `parseSrcComment` is called again, this time with `0:333:444` as input.

Reply:
Ah, the point is that ``Scanner::scanSingleLineDocComment()`` just merges the contents of all consecutive lines starting with ``///`` (but not more than that) - alright!